File: Rails-backend/app/controllers/api/V1/cloud controller.rb
class Api::V1::CloudsController < ApplicationController
  before_action :handle_options_request

  # GET /api/v1/clouds
  def index
    clouds = [
      {
        id: 'aws',
        name: 'Amazon Web Services',
        icon: 'fab fa-aws',
        color: '#FF9900',
        status: 'active'
      },
      {
        id: 'gcp',
        name: 'Google Cloud Platform',
        icon: 'fab fa-google',
        color: '#4285F4',
        status: 'active'
      },
      {
        id: 'azure',
        name: 'Microsoft Azure',
        icon: 'fab fa-microsoft',
        color: '#0078D4',
        status: 'active'
      }
    ]

    render json: { clouds: clouds }
  end

  # GET /api/v1/clouds/services
  def services
    cloud = params[:cloud] || 'aws'
    
    services = case cloud
    when 'aws'
      [
        { name: 'EMR', status: 'running', health: 'healthy' },
        { name: 'S3', status: 'running', health: 'healthy' },
        { name: 'Redshift', status: 'running', health: 'healthy' },
        { name: 'Kinesis', status: 'running', health: 'healthy' },
        { name: 'Lambda', status: 'running', health: 'healthy' }
      ]
    when 'gcp'
      [
        { name: 'Dataproc', status: 'running', health: 'healthy' },
        { name: 'BigQuery', status: 'running', health: 'healthy' },
        { name: 'Cloud Storage', status: 'running', health: 'healthy' },
        { name: 'Pub/Sub', status: 'running', health: 'healthy' },
        { name: 'Cloud Functions', status: 'running', health: 'healthy' }
      ]
    when 'azure'
      [
        { name: 'HDInsight', status: 'running', health: 'healthy' },
        { name: 'Data Lake', status: 'running', health: 'healthy' },
        { name: 'Synapse Analytics', status: 'running', health: 'healthy' },
        { name: 'Event Hubs', status: 'running', health: 'healthy' },
        { name: 'Azure Functions', status: 'running', health: 'healthy' }
      ]
    end

    render json: { 
      cloud: cloud,
      services: services,
      timestamp: Time.current
    }
  end

  # GET /api/v1/clouds/status
  def status
    cloud = params[:cloud] || 'aws'
    
    status_data = {
      cloud: cloud,
      overall_status: 'healthy',
      uptime: '99.9%',
      active_clusters: rand(3..8),
      processing_jobs: rand(10..50),
      data_processed_gb: rand(1000..5000),
      cost_today: rand(100..500),
      last_updated: Time.current
    }

    render json: status_data
  end

  # POST /api/v1/clouds/switch
  def switch
    cloud = params[:cloud]
    
    if %w[aws gcp azure].include?(cloud)
      # Simulate cloud switching logic
      Rails.cache.write('current_cloud', cloud, expires_in: 1.hour)
      
      render json: {
        success: true,
        message: "Switched to #{cloud.upcase}",
        cloud: cloud,
        timestamp: Time.current
      }
    else
      render json: {
        success: false,
        message: 'Invalid cloud provider',
        valid_clouds: %w[aws gcp azure]
      }, status: :bad_request
    end
  end

  # GET /api/v1/clouds/cost_analysis
  def cost_analysis
    cloud = params[:cloud] || Rails.cache.read('current_cloud') || 'aws'
    
    costs = case cloud
    when 'aws'
      { hourly: 24.50, daily: 588.00, monthly: 17640.00 }
    when 'gcp'
      { hourly: 22.80, daily: 547.20, monthly: 16416.00 }
    when 'azure'
      { hourly: 26.20, daily: 628.80, monthly: 18864.00 }
    end

    breakdown = {
      compute: costs[:hourly] * 0.6,
      storage: costs[:hourly] * 0.2,
      networking: costs[:hourly] * 0.1,
      other: costs[:hourly] * 0.1
    }

    render json: {
      cloud: cloud,
      costs: costs,
      breakdown: breakdown,
      currency: 'USD',
      timestamp: Time.current
    }
  end
end

File: backend/app/HTTP/controllers/cloud controller.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Request;
use Illuminate\Http\JsonResponse;

class CloudController extends Controller
{
    /**
     * Get cloud services status for a specific platform
     */
    public function getServices(string $platform): JsonResponse
    {
        $services = [
            'aws' => [
                ['name' => 'EMR', 'status' => 'success', 'icon' => 'fas fa-server'],
                ['name' => 'S3', 'status' => 'success', 'icon' => 'fas fa-database'],
                ['name' => 'Redshift', 'status' => 'success', 'icon' => 'fas fa-warehouse'],
                ['name' => 'Kinesis', 'status' => 'warning', 'icon' => 'fas fa-stream'],
                ['name' => 'Lambda', 'status' => 'success', 'icon' => 'fas fa-bolt']
            ],
            'gcp' => [
                ['name' => 'Dataproc', 'status' => 'success', 'icon' => 'fas fa-server'],
                ['name' => 'BigQuery', 'status' => 'success', 'icon' => 'fas fa-database'],
                ['name' => 'Cloud Storage', 'status' => 'success', 'icon' => 'fas fa-cloud'],
                ['name' => 'Pub/Sub', 'status' => 'success', 'icon' => 'fas fa-stream'],
                ['name' => 'Cloud Functions', 'status' => 'warning', 'icon' => 'fas fa-bolt']
            ],
            'azure' => [
                ['name' => 'HDInsight', 'status' => 'success', 'icon' => 'fas fa-server'],
                ['name' => 'Data Lake', 'status' => 'success', 'icon' => 'fas fa-water'],
                ['name' => 'Synapse Analytics', 'status' => 'success', 'icon' => 'fas fa-chart-line'],
                ['name' => 'Event Hubs', 'status' => 'success', 'icon' => 'fas fa-stream'],
                ['name' => 'Azure Functions', 'status' => 'success', 'icon' => 'fas fa-bolt']
            ]
        ];

        return response()->json($services[$platform] ?? []);
    }

    /**
     * Get cost analysis for a specific platform
     */
    public function getCostAnalysis(string $platform): JsonResponse
    {
        $costs = [
            'aws' => ['hourly' => 24.50, 'monthly' => 17640],
            'gcp' => ['hourly' => 22.80, 'monthly' => 16416],
            'azure' => ['hourly' => 26.20, 'monthly' => 18864]
        ];

        return response()->json($costs[$platform] ?? $costs['aws']);
    }

    /**
     * Get cluster status for platform and engine
     */
    public function getClusterStatus(string $platform, string $engine): JsonResponse
    {
        $status = [
            'aws' => [
                'hadoop' => ['status' => 'running', 'nodes' => 5, 'version' => '3.3.4'],
                'spark' => ['status' => 'running', 'nodes' => 8, 'version' => '3.4.1']
            ],
            'gcp' => [
                'hadoop' => ['status' => 'running', 'nodes' => 4, 'version' => '3.3.4'],
                'spark' => ['status' => 'running', 'nodes' => 6, 'version' => '3.4.1']
            ],
            'azure' => [
                'hadoop' => ['status' => 'running', 'nodes' => 6, 'version' => '3.3.4'],
                'spark' => ['status' => 'running', 'nodes' => 10, 'version' => '3.4.1']
            ]
        ];

        return response()->json($status[$platform][$engine] toISOString }
}